### ISSUE TITLE
Flawed dispatch triggered by changing (irrelevant) contents of a for loop over empty collection

### ISSUE THREAD

**Issue by ablaom** [JuliaLang/julia#36907]
After investigating the source of  https://github.com/alan-turing-institute/ScientificTypes.jl/issues/104 I came up with the following MWE which, it appears to me, points to a Julia bug.

I've no idea why a `SentinelArray` triggers the weird dispatch and the other `AbstractArray` examples don't.

So here's the unexpected (I dare say wrong) behaviour. 

```julia
using SentinelArrays, SparseArrays

abstract type IncorrectDispatch end

const dict = Dict{Symbol,Function}()

# as `dict` is empty the contents of the `for` block shouldn't matter:
function trait(X)::Symbol
    for (name, f) in dict
        if f(X)
            return name
        end
    end
    return :other
end

scitype(X) = scitype(X, Val(trait(X)))
scitype(X, ::Val{:other}) = IncorrectDispatch

# overload for AbstractArrays:
scitype(A::AbstractArray{T}, ::Val{:other}) where T =
    "correct dispatch"

# three different AbstractArrays with same elements:
a = [1, 2, 3, 4]
b = sparse(a)
c = SentinelArray(a)

# The `trait` function appears to work:
@assert trait(a) == :other
@assert trait(b) == :other
@assert trait(c) == :other

# these should all return  "correct dispatch" but the last one does not:
julia> scitype(a)
"correct dispatch"

julia> scitype(b)
"correct dispatch"

julia> scitype(c)
IncorrectDispatch
```

If, however, I change the contents of the for loop (which ought never to be executed) the wrong behaviour goes away:

```julia
function trait(X)::Symbol
    for (name, f) in dict
        if f(X)
            return :other
        end
    end
    return :other
end

julia> scitype(c)
"correct dispatch"
```

```julia
julia> Pkg.installed()
┌ Warning: Pkg.installed() is deprecated
└ @ Pkg /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.4/Pkg/src/Pkg.jl:531
Dict{String,VersionNumber} with 3 entries:
  "ScientificTypes"    => v"0.8.1"

julia> versioninfo()
Julia Version 1.4.2
Commit 44fa15b150* (2020-05-23 18:35 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
Environment:
  JULIA_PATH = /Applications/Julia-1.4.app/Contents/Resources/julia/bin/julia
  JULIA_NUM_THREADS = 5
```

**Comment by Keno** [comment]
SentinelArrays currently triggers some bad code paths in the type system (https://github.com/JuliaLang/julia/issues/36869). Probably worth waiting for that to be fixing before trying to reduce this much farther, since bad type system behavior can have all sorts of unexpected impacts.

**Comment by JeffBezanson** [comment]
Confirmed fixed by #36996
